File: quickstart_guides/arrays_strings/python/find_substring.py
"""
Title: Find substring

Problem:
    You are given a string, s, and a list of words, words, that are all of the
    same length. Find all starting indices of substring(s) in s that is a
    concatenation of each word in words exactly once and without any
    intervening characters.

Execution: python find_substring.py
"""
from collections import Counter
from typing import List
import unittest


def find_substring(s: str, words: List[str]) -> List[int]:
    if not words or not s:
        return []

    wordlen, nwords = len(words[0]), len(words)
    indices = []

    # iterative over `wordlen` times
    for i in range(wordlen):
        counter = Counter(words)
                    # start index of sliding window left and right pointer
        l = r = i
        while r + wordlen <= len(s):
            # remove the substring from the counter
            counter[s[r:r+wordlen]] -= 1
            r += wordlen
            # move left pointer one word offset to the right once the window size exceeds
            if r - l > wordlen * nwords:
                counter[s[l:l+wordlen]] += 1
                l += wordlen
            # if the window size is what we want and we found that all values in counter are 0
            # we can make sure that the substring in this window is one of the answer
            # we are looking for
            if r - l == wordlen * nwords and all(count == 0 for count in counter.values()):
                indices.append(l)

    return indices

class TestFindSubstring(unittest.TestCase):
    """Unit tests for find_substring."""

    def test_1(self):
        self.assertEqual(find_substring("barfoothefoobarman", ["foo", "bar"]), [0, 9])

    def test_2(self):
        self.assertEqual(find_substring("wordgoodgoodgoodbestword", ["word", "good", "best", "word"]), [])


if __name__ == '__main__':
    unittest.main()

File: quickstart_guides/arrays_strings/java/FindSubstring.java
/*
 *   Title: Find substring
 *
 *   Problem:
        You are given a string, s, and a list of words, words, that are all of the
        same length. Find all starting indices of substring(s) in s that is a
        concatenation of each word in words exactly once and without any
        intervening characters.
 *
 *   Execution: javac FindSubstring.java && java FindSubstring
 */

import java.util.*;

public class FindSubstring {
    public static List<Integer> findSubstring(String s, String[] words) {
        List<Integer> result = new ArrayList<Integer>();

        if(s.length() == 0 || words.length == 0) return result;
        int length = words[0].length();
        int sLength = s.length();

        HashMap<String,Integer> concat = new HashMap<String,Integer>();

        for(String word : words){
            concat.put(word,concat.getOrDefault(word,0)+1);
        }
        HashMap<String,Integer> local = new HashMap<String,Integer>();

        int start = 0;
        int end = length-1;

        while(end < sLength){
            //System.out.println(end);
            String sub = s.substring(end-length+1,end+1);
            if(concat.containsKey(sub)){
                local.put(sub,local.getOrDefault(sub,0)+1);
                if(concat.get(sub) == 1) concat.remove(sub);
                else concat.put(sub,concat.get(sub) - 1);
                end += length;
            }else {

                if(!local.isEmpty()){
                    for(Map.Entry<String,Integer> entry : local.entrySet()){
                        String word = entry.getKey();
                        int value = entry.getValue();
                        concat.put(word,concat.getOrDefault(word,0)+value);
                    }
                    local = new HashMap<>();
                }
                start++;
                end = start + length - 1;
            }
            if(concat.isEmpty()){
                result.add(start);
                concat = local;
                local = new HashMap<String,Integer>();
                start++;
                end = start + length -1;
            }
        }

        return result;
    }

    public static void main(String[] args) {

        List<Integer> expected_output_1 = new ArrayList<Integer>();
        expected_output_1.add(0);
        expected_output_1.add(9);

        String[] words_test_1 = {"foo", "bar"};

        assert findSubstring("barfoothefoobarman", words_test_1) == expected_output_1;
        System.out.println("Passed all test cases");
    }
}
